### ISSUE TITLE
Hard (static) dependency on harfbuzz broke JavaFX on GraalVM

### ISSUE THREAD

**Issue by johanvos** [oracle/graal#3213]
**Describe the issue**

https://github.com/oracle/graal/commit/cb492e6b8cc099ea1ed0161504d0092246fd2649 broke JavaFX applications

This line is introduced:

`nativeLibraries.addStaticJniLibrary("harfbuzz");`

This breaks applications that have native code that is *dynamically* linked with harfbuzz (e.g. JavaFX)

example stacktrace:
```

#0  0x00007fffd04a21d2 in  ()
#1  0x00007ffff63dedeb in hb_face_reference_table ()
    at /lib/x86_64-linux-gnu/libharfbuzz.so.0
#2  0x00007ffff640d4bd in  () at /lib/x86_64-linux-gnu/libharfbuzz.so.0
#3  0x00007ffff645366c in hb_ot_metrics_get_position ()
    at /lib/x86_64-linux-gnu/libharfbuzz.so.0
#4  0x00007ffff6d36d00 in pango_fc_font_create_base_metrics_for_context ()
    at /lib/x86_64-linux-gnu/libpangoft2-1.0.so.0
#5  0x00007ffff669bf58 in  () at /lib/x86_64-linux-gnu/libpangocairo-1.0.so.0
#6  0x00007ffff72e0ecf in  () at /lib/x86_64-linux-gnu/libpango-1.0.so.0
#7  0x00007ffff72e1209 in  () at /lib/x86_64-linux-gnu/libpango-1.0.so.0
#8  0x00007ffff72e14da in  () at /lib/x86_64-linux-gnu/libpango-1.0.so.0
#9  0x00007ffff72e5fa8 in  () at /lib/x86_64-linux-gnu/libpango-1.0.so.0
#10 0x00007ffff76489c9 in  () at /lib/x86_64-linux-gnu/libgtk-3.so.0
```

**Comment by oubidar-Abderrahim** [comment]
The issue is tracked internally on GR-29386

**Comment by gradinac** [comment]
After investigating this a bit further, JavaFX builds do not seem to be affected by:
```
nativeLibraries.addStaticJniLibrary("harfbuzz");
```

The core issue seems to be the bump of the JDK version from 21.0-b03 to 21.0-b06. Builidng GraalVM master with 21.0-b03 and building and running the reproducer from https://github.com/gluonhq/client-maven-plugin/issues/299#issuecomment-778100893 works. However, when building GraalVM with 21.0-b06, the reproducer segfaults both with and without this library being added in native-image.

One notable change from 21.0-b03 to 21.0-b06 is that `harfbuzz` is no longer part of `fontmanager` - rather, it is a standalone library, with both static and dynamic libraries available in the JDK (with the static one only being used by native-image). Does `substrate` dynamically link against `harfbuzz` at some point? If that's the case, and the path to the JDK dynamic libraries is on the linker path, the final executable could link against the JDK's `harfbuzz` (instead of the one from the system) - if the two are incompatible, that could be the source of these issues.

**Comment by jerboaa** [comment]
Wouldn't the same happen if the application (or some of it's deps) dynamically link to system libharfbuzz.so yet native-image detecting the harfbuzz dep from the JDK (e.g. because of some font shaping). Then if the JDK's harfbuzz is not compatible with the system libharfbuzz the issue is being triggered. I wonder if `-H:NativeLinkerOptions=-Wl,-whole-archive` would shed any light on this.

**Comment by arodionov** [comment]
> Does substrate dynamically link against harfbuzz at some point?

I think this can be checked by temporarily deleting or renaming this lib in JDK.
Maybe `ldd` can be also helpful.

**Comment by johanvos** [comment]
Substrate links with the system libharfbuzz.so (not with the one in the JDK lib directory)
@jerboaa That is my thinking too: native-image somehow "includes" parts of the JDK's harfbuzz, while we link the final image with the (dynamic) system libharfbuzz.so.
We don't use the GraalVM native-image linker, but if I link harfbuzz with `-Wl,--whole-archive` I get the same error. 

However, if I don't link harfbuzz at all, there is no crash.

I added a breakpoint to the broken function (`hb_face_reference_table`) and that wasn't defined in the image. The bp was activated when the path-to-crash was triggered, since the system libharfbuzz.so was loaded. However, this time there was no crash. From the disassembled code, it seems to me that when there is a crash`hb_face_reference_table` was called with a `null` argument.

**Comment by johanvos** [comment]
After some more digging, thanks to @gradinac it is now clear where the problem originates from.

We traced this down to (as Aleksandar assumed) the libharfbuzz.a from the JDK distribution being linked at linktime (it is higher on the librarypath because the JDK libs dir is there because we need to link with the java libs), while the system libharfbuzz.so (/usr/lib...) being used at runtime.
Although we can work around this, I think in general this is a dangerous pattern. If the JDK library path is somehow on LD_LIBRARY_PATH (e.g. whenever you want to have the java libs linked in an image), the libraries there take precedence over the system libraries if they have the same name. In case of e.g. "libjava.a" that is exactly what we want, but in case of harfbuzz, this is not what we want.

In general, I would think it's better to use e.g. libharfbuzzjdk.a or similar, to clearly distinguish between the distribution-specified library and the jdk-specific one. 

We saw a related problem in the past, when creating static images (with GraalVM native-image) using JavaFX and OpenCV.  JavaFX has the libjpeg sources into the OpenJFX repository, and the symbols are part of libjavafx_iio. However, since OpenCV has native libs that depend on libjpeg.so, we have duplicate symbols. The names of the libraries are different, so the problem is not exactly the same as this one, but I believe the recommended solution is related: if we include external code (copy-by-value) somewhere, it is better to prefix the public symbols, and to rename the library name. That would avoid conflicts between libraries with the same name, or libraries with different names but similar symbols.

In summary, this rather seems a JDK issue than a GraalVM issue, so I think we can close this.

**Comment by jerboaa** [comment]
Did anyone open an OpenJDK issue for this?

**Comment by johanvos** [comment]
@jerboaa  I created https://bugs.openjdk.java.net/browse/JDK-8250894 to track this.

**Comment by jerboaa** [comment]
@johanvos Are you sure that's the right bug? It seems an old bug which adds the possibility to use system libharfbuzz. It doesn't seem to address the name of the OpenJDK-included harfbuzz library.

Note that JDK-8250894 shipped with the latest JDK 11u update and somewhat triggered https://github.com/oracle/graal/commit/cb492e6b8cc099ea1ed0161504d0092246fd2649

**Comment by oubidar-Abderrahim** [comment]
If this issue is still relevant in the latest GraalVM versions, please create a new issue following the template.
